src/Session.h

```diff
@@ -0,0 +1,301 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+
+#pragma once
+
+#include <memory>
+
+#include "zeek/Hash.h"
+#include "zeek/Obj.h"
+#include "zeek/EventHandler.h"
+#include "zeek/Timer.h"
+
+namespace zeek {
+
+class RecordVal;
+using RecordValPtr = IntrusivePtr<RecordVal>;
+class Session;
+
+namespace analyzer { class Analyzer; }
+namespace detail {
+	class SessionTimer;
+	class SessionKey;
+}
+
+typedef void (Session::*timer_func)(double t);
+
+class Session : public Obj {
+public:
+
+	/**
+	 * Construct a new session.
+	 *
+	 * @param t The timestamp for the start and end of the session when it's initially
+	 * created. The end time will be updated as later packets are processed.
+	 * @param timeout_event The event that will be emitted when the session times out.
+	 * @param status_update_event The event that will be emitted for session status
+	 * updates. This can be set to null to disable status updates. This event also
+	 * won't be emitted unless the EnableStatusUpdateTimer() method is called on the
+	 * session, and the interval is set to a non-zero value.
+	 * @param status_update_interval The interval in seconds for the status update
+	 * event to be emitted. Setting this to zero disables the status update timer.
+	 */
+	Session(double t, EventHandlerPtr timeout_event,
+	        EventHandlerPtr status_update_event = nullptr,
+	        double status_update_interval = 0);
+
+	virtual ~Session() {}
+
+	/**
+	 * Invoked when the session is about to be removed. Use Ref(this)
+	 * inside Done to keep the session object around, though it'll
+	 * no longer be accessible from the SessionManager.
+	 */
+	virtual void Done() = 0;
+
+	/**
+	 * Returns a key for the session. This is used as the key for storing
+	 * the session in SessionManager.
+	 *
+	 * @param copy Flag to indicate that the key returned must have a copy of the
+	 * key data instead of just a pointer to it.
+	 */
+	virtual detail::SessionKey SessionKey(bool copy) const = 0;
+
+	/**
+	 * Set the key as invalid.
+	 */
+	void ClearKey()	{ key_valid = true; }
+
+	/**
+	 * Return whether the key is valid.
+	 */
+	bool IsKeyValid() const	{ return key_valid; }
```

Comment:
What do you think about now renaming these to something like `IsInSessionTable` and `SetInSessionTable` (or so)? The notion of a key being valid or not doesn't really seem to be relevant for the API anymore.

Reply:
Seems reasonable to me.